### PR TITLE
implement services for dynamic data

### PR DESCRIPTION
**Pull request by dirk-thomas**
Connects to ros2/ros2#27

@esteve @tfoote @wjwwood Please review.

### CODE REVIEW

**Comment by tfoote** [comment]
lgtm

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp [hunk 1]
```diff
@@ -907,18 +1172,98 @@ rmw_create_service(
...
+  DDSDataReader * request_datareader = replier->get_request_datareader();
+
+  ConnextDynamicServiceInfo * server_info = new ConnextDynamicServiceInfo();
+  server_info->replier_ = replier;
+  server_info->request_datareader_ = request_datareader;
+  server_info->response_type_support_ = response_type_support;
+  server_info->request_members_ = request_members;
+  server_info->response_members_ = response_members;
+
+  rmw_service_t * service = new rmw_service_t;
```
Comment (esteve) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: Please add a comment about using the allocator functions in the future.

**Comment by esteve** [comment]
+1

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp (hunk 1 above)
Comment (dirk-thomas) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: Added in f8f5e8be2bf0d004a10ad729b1a4235f056465b0 - also for the static Connext implementation.

**Comment by dirk-thomas** [comment]
Passing test: http://54.183.26.131:8080/view/ros2/job/ros2_batch_ci_linux/62/testReport/(root)/test_request_response__rmw_connext_xtypes_dynamic_cpp/

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp [hunk 2]
```diff
@@ -889,16 +1095,76 @@ rmw_take_response(
...
+    return RMW_RET_ERROR;
+  }
+
+  ConnextDynamicClientInfo * client_info = \
+    static_cast<ConnextDynamicClientInfo *>(client->data);
+  connext::Requester<DDS_DynamicData, DDS_DynamicData> * requester = client_info->requester_;
+
+  connext::LoanedSamples<DDS::DynamicData> replies = requester->take_replies(1);
+  typedef connext::LoanedSamples<DDS::DynamicData>::iterator iterator;
+  for (iterator it = replies.begin(); it != replies.end(); ++it) {
```
Comment (wjwwood) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: You could use the for range loop or even just `auto it = replies.begin()`. It's not important to change this one, but we are using C++11, we should make the most of it!

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp [hunk 3]
```diff
@@ -889,16 +1095,76 @@ rmw_take_response(
...
+    printf("  rmw_take_response() got a reply\n");
```
Comment (wjwwood) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: Also, should we leave this kind of thing in? (underscores the need for a proper logging system)

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp [hunk 4]
```diff
@@ -889,16 +1095,76 @@ rmw_take_response(
...
+  }
+  if (replies.begin() != replies.end() && replies.begin()->info().valid_data) {
+    _take(&replies.begin()->data(), ros_response, client_info->response_members_);
+
+    rmw_request_id_t & req_id = *(reinterpret_cast<rmw_request_id_t *>(ros_request_header));
+    req_id.sequence_number = ((int64_t)replies.begin()->related_identity().sequence_number.high)
+      << 32 | replies.begin()->related_identity().sequence_number.low;
```
Comment (wjwwood) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: The use of parenthesis here seems strange to me. I would have thought that you would have included the `<< 32` in the parenthesis. Was this intentional?

**Comment by wjwwood** [comment]
Other than the comments, lgtm.

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp (hunk 3 above)
Comment (dirk-thomas) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: This is currently consistent with all the other rmw implementation. But it should for sure be removed at some point.

Comment (dirk-thomas) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: Nvm, this should actually be removed right now.

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp (hunk 4 above)
Comment (dirk-thomas) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: Just copied from the static implementation: https://github.com/ros2/rmw_connext/blob/1aece5657d69a34e8c808943105e3bb45a66c2ad/rosidl_typesupport_connext_cpp/resource/srv_ServiceTypeSupport.cpp.template#L126

I will wrap it in parenthesis in both locations.

File: rmw_connext_xtypes_dynamic_cpp/src/functions.cpp (hunk 2 above)
Comment (dirk-thomas) [inline rmw_connext_xtypes_dynamic_cpp/src/functions.cpp]: Gone.

**Pull request merged**